File: solutions/typescript/matching-brackets/1/matching-brackets.ts
export function isPaired(input: string): boolean{
  let stack = [];
  let openBrackets = ['{', '[', '('];
  let closeBrackets = ['}', ']', ')'];
  let bracketPairs: { [key: string]: string } = {
    '}': '{',
    ']': '[',
    ')': '('
  }
  while (input.length > 0){
    let char = input[0];
    input = input.substring(1);
    if (openBrackets.includes(char)){
      stack.push(char);
    } else if (closeBrackets.includes(char)){
      if (stack.length === 0 || stack[stack.length - 1] !== bracketPairs[char]){
        return false;
      }
      stack.pop();
    }
  }
  return stack.length === 0;
}

File: solutions/c/matching-brackets/1/matching_brackets.c
#include "matching_brackets.h"
#include <string.h>
#include <stdlib.h>

bool is_paired(const char *input) {
    char stack[strlen(input)];
    size_t top = 0;

    for (size_t i = 0; i < strlen(input); i++) {
        char c = input[i];
        if (c == '(' || c == '[' || c == '{') {
            stack[top++] = c;
        } else if (c == ')' || c == ']' || c == '}') {
            if (top == 0) {
                return false;
            }
            char last = stack[top - 1];
            if (last == '(' && c == ')') {
                top--;
            } else if (last == '[' && c == ']') {
                top--;
            } else if (last == '{' && c == '}') {
                top--;
            } else {
                return false;
            }
        }
    }
    return top == 0;
}
